{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2: Define Connection Masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Connectivity Matrices\n",
    "# The connections between the layers are defined by the connectivity matrix.\n",
    "\n",
    "# The matrix has the shape of (N_nodes_layer_1, N_nodes_layer_2).\n",
    "# It is a sparse matrix with zeros for no connections and ones if there is a connections. For example.\n",
    "\n",
    "\n",
    "#             output\n",
    "#           1 2 3 4 5\n",
    "# input 1 | 1 0 0 0 0 |\n",
    "# input 2 | 1 1 0 0 0 |\n",
    "# input 3 | 0 1 0 0 0 |\n",
    "# input 4 | 0 1 0 0 0 |\n",
    "# input 5 | 0 0 1 0 0 |\n",
    "# input 6 | 0 0 0 1 0 |\n",
    "# input 7 | 0 0 0 1 0 |\n",
    "\n",
    "\n",
    "# This connects the first two inputs (1,2) to the first neuron in the second layer.\n",
    "# Connects input 2,3 and 4 to output neuron 2.\n",
    "# Connects input 5 to output neuron 3\n",
    "# Connects input 6 and 7 o the 4th neuron in the subsequent layer\n",
    "# Connects nothing to the 5th neuron\n",
    "\n",
    "#imports & paths\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import scipy\n",
    "import h5py\n",
    "import tables\n",
    "from scipy import stats\n",
    "import glob\n",
    "import itertools\n",
    "import tables\n",
    "import tqdm\n",
    "\n",
    "basepath = os.getcwd()[:-4]\n",
    "hasepath = basepath + \"/hase/\"\n",
    "savepath = basepath + \"/processed_data/\"\n",
    "studyname = str(np.load(savepath + \"studyname.npy\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate Input files for Annovar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(823497, 6)\n",
      "Number of variants (8590, 2)\n",
      "(8590, 2)\n",
      "(823497, 6)\n",
      "(823497, 6)\n",
      "/media/avanhilten/pSSD450/GenNet//processed_data//annovar_input_BulgarianTrio.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp</th>\n",
       "      <th>bp2</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>index_col</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>15903</td>\n",
       "      <td>15903</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>63735</td>\n",
       "      <td>63735</td>\n",
       "      <td>C</td>\n",
       "      <td>CCTA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>120994</td>\n",
       "      <td>120994</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>267227</td>\n",
       "      <td>267227</td>\n",
       "      <td>T</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>267300</td>\n",
       "      <td>267300</td>\n",
       "      <td>TTA</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR      bp     bp2 allele1 allele2  index_col\n",
       "0    1   15903   15903      GC       G          0\n",
       "1    1   63735   63735       C    CCTA          1\n",
       "2    1  120994  120994       A     AAT          2\n",
       "3    1  267227  267227       T    TTAA          3\n",
       "4    1  267300  267300     TTA       T          4"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probes = pd.read_hdf(hasepath + '/probes/'+studyname+'.h5')\n",
    "print(probes.shape)\n",
    "probes.head()\n",
    "\n",
    "if os.path.exists(hasepath +'/probes/'+studyname+'_hash_table.csv.gz'):\n",
    "    hashtable = pd.read_csv(hasepath +'/probes/'+studyname+'_hash_table.csv.gz', compression=\"gzip\", sep='\\t')\n",
    "else: \n",
    "    hashtable = pd.read_csv(hasepath +'/probes/'+studyname+'_hash_table.csv', sep='\\t')\n",
    "\n",
    "\n",
    "\n",
    "hashtable['allele1']  = hashtable['keys']\n",
    "unhashed_probes = probes.merge(hashtable, on='allele1', how = \"left\" )\n",
    "unhashed_probes = unhashed_probes.drop(columns=[\"keys\", \"allele1\"])\n",
    "unhashed_probes = unhashed_probes.rename(columns = {'allele':'allele1'})\n",
    "\n",
    "#reload hashtable for other allele\n",
    "\n",
    "if os.path.exists(hasepath +'/probes/'+studyname+'_hash_table.csv.gz'):\n",
    "    hashtable = pd.read_csv(hasepath +'/probes/'+studyname+'_hash_table.csv.gz', compression=\"gzip\", sep='\\t')\n",
    "else: \n",
    "    hashtable = pd.read_csv(hasepath +'/probes/'+studyname+'_hash_table.csv', sep='\\t')\n",
    "\n",
    "hashtable['allele2']  = hashtable['keys']\n",
    "unhashed_probes = unhashed_probes.merge(hashtable, on='allele2', how = \"left\")\n",
    "unhashed_probes = unhashed_probes.drop(columns=[\"keys\", \"allele2\"])\n",
    "unhashed_probes = unhashed_probes.rename(columns = {'allele':'allele2'})\n",
    "\n",
    "#clean up\n",
    "annovar_input = unhashed_probes.drop(columns=[\"ID\",\"distance\"])\n",
    "annovar_input[\"bp2\"] = annovar_input[\"bp\"]\n",
    "annovar_input[\"index_col\"] = annovar_input.index\n",
    "annovar_input = annovar_input[['CHR', 'bp',\"bp2\",\"allele1\",\"allele2\",\"index_col\"]]\n",
    "\n",
    "print('Number of variants',annovar_input.shape)\n",
    "\n",
    "annovar_input_path = savepath + '/annovar_input_'+studyname+'.csv'\n",
    "annovar_input.to_csv(annovar_input_path,sep=\"\\t\", index=False, header = False)\n",
    "annovar_input.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Annovar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Navigate to annovar (cd /home/avanhilten/annovar/)\n",
      "Run: perl annotate_variation.pl -geneanno -dbtype refGene -buildver hg19 /media/avanhilten/pSSD450/GenNet//processed_data//annovar_input_BulgarianTrio.csv humandb --outfile /media/avanhilten/pSSD450/GenNet//processed_data//BulgarianTrio_RefGene\n"
     ]
    }
   ],
   "source": [
    "print(\"install annovar: https://doc-openbio.readthedocs.io/projects/annovar/en/latest/user-guide/download/\")\n",
    "print(\"Navigate to annovar, e.g cd /home/charlesdarwin/annovar/\")\n",
    "print(\"Update annovar: annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene humandb/\")\n",
    "print(\"Run: perl annotate_variation.pl -geneanno -dbtype refGene -buildver hg19 \"+str(savepath)+\"/annovar_input_\"+str(studyname)+\".csv humandb --outfile \"+str(savepath)+\"/\"+str(studyname)+\"_RefGene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create mask from gene annotations\n",
    "\n",
    "Here we create the mask for the gene layer. Note the layer here assigns all gees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(800465, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>FAM138A(dist=27654),OR4F5(dist=5356)</td>\n",
       "      <td>1</td>\n",
       "      <td>63735</td>\n",
       "      <td>63735</td>\n",
       "      <td>C</td>\n",
       "      <td>CCTA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>OR4F5(dist=50986),LOC729737(dist=13779)</td>\n",
       "      <td>1</td>\n",
       "      <td>120994</td>\n",
       "      <td>120994</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC729737(dist=126661),LOC100132287(dist=56665)</td>\n",
       "      <td>1</td>\n",
       "      <td>267227</td>\n",
       "      <td>267227</td>\n",
       "      <td>T</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC100288069(dist=3358),FAM87B(dist=35325)</td>\n",
       "      <td>1</td>\n",
       "      <td>717426</td>\n",
       "      <td>717426</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC100288069(dist=31409),FAM87B(dist=7274)</td>\n",
       "      <td>1</td>\n",
       "      <td>745477</td>\n",
       "      <td>745477</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                                                1  2       3  \\\n",
       "0  intergenic             FAM138A(dist=27654),OR4F5(dist=5356)  1   63735   \n",
       "1  intergenic          OR4F5(dist=50986),LOC729737(dist=13779)  1  120994   \n",
       "2  intergenic  LOC729737(dist=126661),LOC100132287(dist=56665)  1  267227   \n",
       "3  intergenic       LOC100288069(dist=3358),FAM87B(dist=35325)  1  717426   \n",
       "4  intergenic       LOC100288069(dist=31409),FAM87B(dist=7274)  1  745477   \n",
       "\n",
       "        4  5     6  7  \n",
       "0   63735  C  CCTA  1  \n",
       "1  120994  A   AAT  2  \n",
       "2  267227  T  TTAA  3  \n",
       "3  717426  C    CT  5  \n",
       "4  745477  T     C  6  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation = pd.read_csv(savepath +str(studyname)+\"_RefGene.variant_function\",sep='\\t', header=None)\n",
    "print(gene_annotation.shape)\n",
    "gene_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Nulls 23032\n",
      "annotated shape: (823497,)\n",
      "number of unique genes: 84483\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp</th>\n",
       "      <th>bp2</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>index_col</th>\n",
       "      <th>into/exonic</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>15903</td>\n",
       "      <td>15903</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>63735</td>\n",
       "      <td>63735</td>\n",
       "      <td>C</td>\n",
       "      <td>CCTA</td>\n",
       "      <td>1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>120994</td>\n",
       "      <td>120994</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>267227</td>\n",
       "      <td>267227</td>\n",
       "      <td>T</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>3</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC729737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>267300</td>\n",
       "      <td>267300</td>\n",
       "      <td>TTA</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR      bp     bp2 allele1 allele2  index_col into/exonic       gene\n",
       "0    1   15903   15903      GC       G          0         NaN        NaN\n",
       "1    1   63735   63735       C    CCTA          1  intergenic    FAM138A\n",
       "2    1  120994  120994       A     AAT          2  intergenic      OR4F5\n",
       "3    1  267227  267227       T    TTAA          3  intergenic  LOC729737\n",
       "4    1  267300  267300     TTA       T          4         NaN        NaN"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns = ['into/exonic', 'gene', 'chr', 'bps', 'bpe', \"mutation1\" ,\"mutation2\" ,'index_col']\n",
    "annovar_annotated = annovar_input.merge(gene_annotation[['into/exonic', 'gene',\"index_col\"]], on='index_col', how = \"left\")\n",
    "print(\"Number of Nulls\",annovar_annotated[\"gene\"].isnull().sum())\n",
    "print(\"annotated shape:\",annovar_annotated[\"gene\"].shape)\n",
    "print(\"number of unique genes:\",len(gene_annotation[\"gene\"].unique()))\n",
    "annovar_annotated['gene'] = annovar_annotated['gene'].str.replace(r\"\\(.*\\)\",\"\",)\n",
    "annovar_annotated['gene'] = annovar_annotated['gene'].str.replace(r\"\\,.*\",\"\",)\n",
    "annovar_annotated.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(754132, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp</th>\n",
       "      <th>bp2</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>index_col</th>\n",
       "      <th>into/exonic</th>\n",
       "      <th>gene</th>\n",
       "      <th>snp_std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>63735</td>\n",
       "      <td>63735</td>\n",
       "      <td>C</td>\n",
       "      <td>CCTA</td>\n",
       "      <td>1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.352587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>120994</td>\n",
       "      <td>120994</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.169120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>267227</td>\n",
       "      <td>267227</td>\n",
       "      <td>T</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>3</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC729737</td>\n",
       "      <td>0.114758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>717426</td>\n",
       "      <td>717426</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>5</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC100288069</td>\n",
       "      <td>0.032759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>745477</td>\n",
       "      <td>745477</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LOC100288069</td>\n",
       "      <td>0.040230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR      bp     bp2 allele1 allele2  index_col into/exonic          gene  \\\n",
       "1    1   63735   63735       C    CCTA          1  intergenic       FAM138A   \n",
       "2    1  120994  120994       A     AAT          2  intergenic         OR4F5   \n",
       "3    1  267227  267227       T    TTAA          3  intergenic     LOC729737   \n",
       "5    1  717426  717426       C      CT          5  intergenic  LOC100288069   \n",
       "6    1  745477  745477       T       C          6  intergenic  LOC100288069   \n",
       "\n",
       "    snp_std  \n",
       "1  0.352587  \n",
       "2  0.169120  \n",
       "3  0.114758  \n",
       "5  0.032759  \n",
       "6  0.040230  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select annotated, autosomal and SNPs with std > 0. Adjust here if you want to use other criteria\n",
    "annovar_annotated[\"snp_std\"] = np.load(savepath + studyname + '_std.npy')\n",
    "annovar_annotated  = annovar_annotated[(annovar_annotated['gene'] != \"NONE\")\n",
    "                                       & (annovar_annotated['CHR'] < 23)\n",
    "                                       & (annovar_annotated['snp_std'] > 0)]\n",
    "annovar_annotated = annovar_annotated.dropna()\n",
    "print(annovar_annotated.shape)\n",
    "annovar_annotated.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/avanhilten/RFreg3/lib/python3.5/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>bp</th>\n",
       "      <th>bp2</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>index_col</th>\n",
       "      <th>into/exonic</th>\n",
       "      <th>gene</th>\n",
       "      <th>snp_std</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tot_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>754130</th>\n",
       "      <td>22</td>\n",
       "      <td>51215123</td>\n",
       "      <td>51215123</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>804618</td>\n",
       "      <td>exonic</td>\n",
       "      <td>RABL2B</td>\n",
       "      <td>0.100819</td>\n",
       "      <td>21463</td>\n",
       "      <td>754127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754131</th>\n",
       "      <td>22</td>\n",
       "      <td>51215124</td>\n",
       "      <td>51215124</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>804619</td>\n",
       "      <td>exonic</td>\n",
       "      <td>RABL2B</td>\n",
       "      <td>0.023271</td>\n",
       "      <td>21463</td>\n",
       "      <td>754128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754127</th>\n",
       "      <td>22</td>\n",
       "      <td>51216327</td>\n",
       "      <td>51216327</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>804620</td>\n",
       "      <td>ncRNA_intronic</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>0.046467</td>\n",
       "      <td>21462</td>\n",
       "      <td>754129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754128</th>\n",
       "      <td>22</td>\n",
       "      <td>51216332</td>\n",
       "      <td>51216332</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>804621</td>\n",
       "      <td>ncRNA_intronic</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>0.045715</td>\n",
       "      <td>21462</td>\n",
       "      <td>754130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754129</th>\n",
       "      <td>22</td>\n",
       "      <td>51216453</td>\n",
       "      <td>51216453</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>804622</td>\n",
       "      <td>ncRNA_intronic</td>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>0.032749</td>\n",
       "      <td>21462</td>\n",
       "      <td>754131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR        bp       bp2 allele1 allele2  index_col     into/exonic  \\\n",
       "754130   22  51215123  51215123       T       C     804618          exonic   \n",
       "754131   22  51215124  51215124       A       G     804619          exonic   \n",
       "754127   22  51216327  51216327       C       T     804620  ncRNA_intronic   \n",
       "754128   22  51216332  51216332       T       G     804621  ncRNA_intronic   \n",
       "754129   22  51216453  51216453       T       C     804622  ncRNA_intronic   \n",
       "\n",
       "             gene   snp_std  gene_id  tot_index  \n",
       "754130     RABL2B  0.100819    21463     754127  \n",
       "754131     RABL2B  0.023271    21463     754128  \n",
       "754127  RPL23AP82  0.046467    21462     754129  \n",
       "754128  RPL23AP82  0.045715    21462     754130  \n",
       "754129  RPL23AP82  0.032749    21462     754131  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list = annovar_annotated.drop_duplicates(\"gene\")\n",
    "gene_list[\"gene_id\"] = np.arange(len(gene_list))\n",
    "gene_list = gene_list[[\"gene\",\"gene_id\"]]\n",
    "\n",
    "annovar_annotated = annovar_annotated.merge(gene_list, on=\"gene\")\n",
    "annovar_annotated = annovar_annotated.sort_values(by = \"index_col\", ascending = True)\n",
    "annovar_annotated['tot_index'] = np.arange(len(annovar_annotated))\n",
    "annovar_annotated.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      "  0%|          | 0/1826 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Finalizing and transposing data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 18%|█▊        | 322/1826 [11:42<55:58,  2.23s/it]  "
     ]
    }
   ],
   "source": [
    "#Finalize the input by selecting only the relevant inputs and by transposing the data\n",
    "\n",
    "t = tables.open_file(savepath + studyname + '_genotype_imputed.h5', mode='r')\n",
    "data = t.root.data\n",
    "num_pat = data.shape[1]\n",
    "num_feat = data.shape[0]\n",
    "\n",
    "used_indices = np.zeros(num_feat)\n",
    "used_indices[annovar_annotated.index_col.values]  =1 \n",
    "\n",
    "f = tables.open_file(savepath + studyname + '_genotype_processed.h5', mode='w')\n",
    "array_c = f.create_earray(f.root, 'data', tables.IntCol(), (0,num_feat ), expectedrows=num_pat,filters=tables.Filters(complib='zlib', complevel=1))\n",
    "f.close()\n",
    "\n",
    "f = tables.open_file(savepath + studyname + '_genotype_processed.h5', mode='a')\n",
    "print(\"\\n Finalizing and transposing data...\")\n",
    "for pat in tqdm.tqdm(range(num_pat)):\n",
    "    a = np.transpose(data[:,pat])\n",
    "    a=np.reshape(a, (1,-1))\n",
    "    f.root.data.append(a)\n",
    "f.close()\n",
    "t.close()\n",
    "print(\"Completed\", studyname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mask including all variants\n",
    "data = np.ones(len(annovar_annotated), np.bool)\n",
    "coord = ( annovar_annotated[\"tot_index\"].values, annovar_annotated[\"gene_id\"].values )\n",
    "SNP_gene_matrix = scipy.sparse.coo_matrix(((data),coord),  shape = (len(annovar_annotated), annovar_annotated[\"gene_id\"].max()+1 ))\n",
    "scipy.sparse.save_npz(savepath +'/SNP_gene_mask', SNP_gene_matrix)\n",
    "SNP_gene_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mask including only exonic variants\n",
    "annovar_annotated_exonic = annovar_annotated[annovar_annotated['into/exonic'] == \"exonic\"]\n",
    "data = np.ones(len(annovar_annotated_exonic), np.bool)\n",
    "coord = ( annovar_annotated_exonic[\"tot_index\"].values, annovar_annotated_exonic[\"gene_id\"].values )\n",
    "SNP_gene_matrix = scipy.sparse.coo_matrix(((data),coord),  shape = (len(annovar_annotated), annovar_annotated[\"gene_id\"].max()+1 ))\n",
    "scipy.sparse.save_npz(savepath +'/SNP_gene_matrix_exonic_merged', SNP_gene_matrix)\n",
    "SNP_gene_matrix"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
